### ISSUE TITLE
Investigate sync var serialization issue under qthreads

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#9651]
We seem to have an issue where using sync vars serializes execution such that all tasks end up running on one qthread worker.

I believe the issue is around: https://github.com/chapel-lang/chapel/blob/1bab5a891757648fac132de749c83803e436d0e3/third-party/qthread/qthread-src/src/feb.c#L178-L187

Basically, the code that wakes up a task that is now unblocked from a sync var seems to assume a work-stealing scheduler is being used. The nemesis scheduler (the one we default to) doesn't support work-stealing  (and tasks aren't marked with `QTHREAD_UNSTEALABLE`) so all tasks are getting put into the current shepherd, and then never getting redistributed.

For this I want to create spmd versions of stream that will barrier on a sync var. This should allow us to see poor performance on master (should basically be single core perf) and once this issue is fixed achieving performance with the reference stream will give us confidence that execution isn't serialized and that we're maintaining the correct task/numa affinity between sync var sleeps/wake-ups.

To actually fix the issue we just need to add some conditionals to the check that spawns to `waiter->rdata->shepherd_ptr->ready`. We want to spawn to that shepherd if the task can't be stolen or if work stealing isn't supported (because the scheduler doesn't support it, or because work-stealing was disabled at runtime.)

**Comment by ronawho** [comment]
https://chapel-lang.org/CHIUW/2018/rollinger-slides.pdf and https://www.researchgate.net/publication/325385645_Parallel_Sparse_Tensor_Decomposition_in_Chapel have some results that show very poor performance when using syncs as locks. I suspect it's because of this serialization, but I would like to verify once this issue is resolved.

**Comment by ronawho** [comment]
![image](https://user-images.githubusercontent.com/1588337/41006051-c67ed6e4-68d4-11e8-8251-499a010b8820.png)

shows the serialized performance (sync var version gets single-threaded perf) for an spmd stream like:

```chpl 
coforall tid in 0..#numTasks {
  barrier.barrier();
  for i in chunk(1..m, numTasks, tid) do
    A[i] = B[i] + alpha * C[i];
}
```

**Comment by ronawho** [comment]
I believe this has been resolved by #9737. Updated graphs [here](https://chapel-lang.org/perf/chapcs/?startdate=2018/06/02&enddate=2018/06/11&graphs=spmdbarrierstreamperformancegbs)

![image](https://user-images.githubusercontent.com/1588337/41247978-6d41a4f0-6d64-11e8-89d6-582758c1d6a6.png)

**Comment by ronawho** [comment]
This also seems to resolve the [issue I mentioned](https://github.com/chapel-lang/chapel/issues/9651#issuecomment-394500188) with [SPLATT](https://github.com/thomasrolinger/Chapel-SPLATT) performance

SPLATT repro instructions (on a 28-core (56HT) Broadwell Cray XC node using cray-libsci instead of openblas):

```sh
git clone git@github.com:thomasrolinger/Chapel-SPLATT.git
cd Chapel-SPLATT/
wget https://www.dropbox.com/s/60fksn8f4scfspa/YELP.bin?dl=1 -O YELP.bin

export CHPL_TASKS=qthreads
export CHPL_MEM=jemalloc
make clean
make LIBS=

./bin/splatt YELP.bin --rank=35 --seed=123456 --tol=1e-20 --iters=20
```

To test the sync serialization fix, I'm manually reverting https://github.com/thomasrolinger/Chapel-SPLATT/commit/b96fb240d6505ee4c382da13e8e531d719166af5, which switched from syncs to atomics. Here's the diff: [gist](https://gist.github.com/ronawho/f99b3d0495133c6662b4b6554e7a4646#file-spatt-use-sync-md)

Performance results:

| Config           | Total Time |
| ---------------- | ---------- |
| Before Sync Lock | 19.1s      |
| Master Sync Lock |  5.8s      |
| Atomic Lock      |  5.4s      |

So using a sync var as a lock is ~4x faster than before. The sync var lock is still slightly slower than using an atomic lock, but we're no longer seeing the massive performance hit from sync var serialization.
